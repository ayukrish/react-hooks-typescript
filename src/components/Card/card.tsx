import React from 'react';
import styled from 'styled-components';

interface CardProps {
  imgSrc: string;
  heading: string;
  dataObj: object;
}

const CardWrapper = styled.article`
  width: 300px;
  margin-bottom: 0.625rem;
  box-shadow: rgba(0, 0, 0, 0.16) 0px 2px 2px 0px,
    rgba(0, 0, 0, 0.08) 0px 0px 0px 1px;
  border-radius: 0.625rem;
  overflow: hidden;
  background: rgb(51, 51, 51);
`;

const CardHeader = styled.div`
  width: 100%;
  background: black;
  max-width: 300px;
  max-height: 300px;
  position: relative;
`;

const CardInfo = styled.div`
  color: rgb(255, 152, 0);
  padding: 1.25rem;
`;

const CardTitle = styled.div<{ imgSrc: string }>`
  width: 100%;
  opacity: 0.8;
  position: ${(props) => (!props.imgSrc ? 'relative' : 'absolute')};
  bottom: 0px;
  background: ${(props) => (!props.imgSrc ? 'rgb(51, 51, 51)' : 'black')};
  padding: 0.625rem;
  h2 {
    color: white;
    font-size: 1.625rem;
    font-weight: 400;
  }
`;

const CharacterText = styled.div`
  padding: 0.75rem 0px 0.375rem;
  border-bottom: 1px solid rgb(68, 68, 68);

  span {
    width: 100%;
    font-size: 0.9rem;
    font-weight: 200;
    text-align: right;

    &:first-child {
      font-size: 0.7rem;
      font-weight: 400;
      color: rgb(158, 158, 158);
      text-align: left;
    }
  }
`;

const Card = ({ imgSrc = '', heading = '', dataObj = {} }: CardProps) => (
  <CardWrapper className={'flex column'}>
    <CardHeader>
      {imgSrc && <img src={imgSrc} alt={heading} />}
      <CardTitle imgSrc={imgSrc}>
        <h2>{heading}</h2>
      </CardTitle>
    </CardHeader>
    <CardInfo>
      {Object.entries(dataObj).map(([key, value]) => (
        <CharacterText className={'flex align-center space-between'} key={key}>
          <span>{key}</span>
          <span>{value}</span>
        </CharacterText>
      ))}
    </CardInfo>
  </CardWrapper>
);

export default Card;
