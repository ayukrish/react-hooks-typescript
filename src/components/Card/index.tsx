import React from 'react';
import styled, { withTheme } from 'styled-components';

interface ICardProps {
  imgSrc?: string;
  heading: string;
  dataObj: IDataObj;
  theme: { theme: 'light' | 'dark' };
}

interface IDataObj {
  STATUS?: string;
  SPECIES?: string;
  GENDER?: string;
  ORIGIN?: string;
  LOCATION?: string;
  EPISODE?: string;
  'AIR DATE'?: string;
  TYPE?: string;
  DIMENSION?: string;
}

const CardWrapper = styled.article<{ theme: 'light' | 'dark' }>`
  width: 600px;
  max-height: 320px;
  overflow: hidden;
  background: ${(props) =>
    props.theme === 'light' ? '#000' : 'rgb(60, 62, 68)'};
  border-radius: 0.5rem;
  margin: 0.75rem;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 4px 6px -1px,
    rgba(0, 0, 0, 0.06) 0px 2px 4px -1px;
`;

const CardHeader = styled.div`
  width: 100%;
  flex-basis: 30%;
`;

const CardInfo = styled.div`
  color: rgb(255, 152, 0);
  padding: 1.25rem;
  flex-basis: 70%;
`;

const CardTitle = styled.h2`
  color: white;
  font-size: 1.625rem;
  font-weight: 400;
`;

const CharacterText = styled.div<{ theme: 'light' | 'dark' }>`
  padding: 0.75rem 0px 0.375rem;
  border-bottom: 1px solid rgb(68, 68, 68);

  span {
    width: 100%;
    font-size: 0.9rem;
    font-weight: 200;
    text-align: right;

    &:first-child {
      font-size: 0.7rem;
      font-weight: 400;
      color: ${(props) =>
        props.theme === 'light' ? '#fff' : 'rgb(158, 158, 158)'};
      text-align: left;
    }
  }
`;

const Card: React.FunctionComponent<ICardProps> = ({
  imgSrc = '',
  heading = '',
  dataObj = {},
  theme,
}: ICardProps) => (
  <CardWrapper className="flex row" theme={theme.theme}>
    <CardHeader>
      <img src={imgSrc} alt={heading} />
    </CardHeader>
    <CardInfo>
      <CardTitle>{heading}</CardTitle>
      {Object.entries(dataObj).map(([key, value]) => (
        <CharacterText
          className="flex align-center space-between"
          key={key}
          theme={theme.theme}
        >
          <span>{key}</span>
          <span>{value}</span>
        </CharacterText>
      ))}
    </CardInfo>
  </CardWrapper>
);

export default withTheme(Card);
