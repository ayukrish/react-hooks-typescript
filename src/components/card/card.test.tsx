import React from 'react';
import { render, screen } from '@testing-library/react';

// Components
import Card from './index';

// eslint-disable-next-line react/jsx-props-no-spreading
const renderCard = (props?: any) => render(<Card {...props} />);

describe('<Card>', () => {
  it('should render Card in light mode', () => {
    renderCard({
      theme: {
        theme: 'light',
      },
    });
    expect(screen.getByRole('article')).toMatchSnapshot();
  });

  it('should render Card in dark mode', () => {
    renderCard({
      theme: {
        theme: 'dark',
      },
    });
    expect(screen.getByRole('article')).toMatchSnapshot();
  });
});
