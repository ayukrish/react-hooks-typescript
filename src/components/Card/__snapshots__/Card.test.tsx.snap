// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card> should render Card in dark mode 1`] = `
<article
  class="sc-bdfBwQ evGFoE flex row"
>
  <div
    class="sc-gsTCUz cibcoK"
  >
    <img
      alt=""
      src=""
    />
  </div>
  <div
    class="sc-dlfnbm inVonk"
  >
    <h2
      class="sc-hKgILt hyrQdz"
    />
  </div>
</article>
`;

exports[`<Card> should render Card in light mode 1`] = `
<article
  class="sc-bdfBwQ ehcuxJ flex row"
>
  <div
    class="sc-gsTCUz cibcoK"
  >
    <img
      alt=""
      src=""
    />
  </div>
  <div
    class="sc-dlfnbm inVonk"
  >
    <h2
      class="sc-hKgILt hyrQdz"
    />
  </div>
</article>
`;
